aiohttp/web_response.py

```diff
@@ -337,19 +337,27 @@ def _generate_content_type_header(
         if self._payload_writer is not None:
             return self._payload_writer
 
-        await request._prepare_hook(self)
         return await self._start(request)
 
     async def _start(self, request: 'BaseRequest') -> AbstractStreamWriter:
         self._req = request
+        writer = self._payload_writer = request._payload_writer
+
+        await self._prepare_headers(request, writer)
+        await request._prepare_hook(self)
+        await self._write_headers(request, writer)
+
+        return writer
 
+    async def _prepare_headers(
+        self, request: 'BaseRequest', writer: AbstractStreamWriter
```

Comment:
Should it accept a `writer` argument given that the same thing is effectively available via `self._payload_writer` anyway?

Reply:
I can sure change it.

Reply:
Makes sense. `request` is available as `self._req` also.

Reply:
The issue is `self._payload_writer` and `self._req` are `Optional[...]` so `_prepare_headers` and `_write_headers` would need to guard against them being `None` to pacify mypy. Should they raise errors if that happens? Passing the non-Optional values might be preferable.